lib/elixir/lib/calendar/iso.ex

```diff
@@ -663,6 +664,147 @@ defmodule Calendar.ISO do
     end
   end
 
+  @doc """
+  Parses an ISO 8601 formatted duration string to a list of `Duration` compabitble unit pairs.
+
+  """
+  @doc since: "1.17.0"
+  @spec parse_duration(String.t()) :: {:ok, [Duration.unit_pair()]} | {:error, atom()}
+  def parse_duration("P" <> string) do
+    parse_date_duration(string, %{}, "")
+  end
+
+  def parse_duration(_) do
+    {:error, :invalid_duration}
+  end
+
+  defp parse_date_duration(_, {:error, error}, _), do: {:error, error}
+
+  defp parse_date_duration(<<>>, duration, ""), do: {:ok, Keyword.new(duration)}
+
+  defp parse_date_duration(<<c, rest::binary>>, duration, buffer)
+       when c in ?0..?9 or c in [?., ?-] do
+    parse_date_duration(rest, duration, <<buffer::binary, c>>)
+  end
+
+  defp parse_date_duration(<<"Y", rest::binary>>, duration, buffer) do
+    duration = parse_buffer_duration(:year, duration, buffer)
+    parse_date_duration(rest, duration, "")
```

Comment:
```suggestion
    with %{} = duration <- parse_buffer_duration(:year, duration, buffer) do
      parse_date_duration(rest, duration, "")
    end
```

Another option is to have a look-ahead. See if the value immediately after "Y" is a digit:

```
parse_date_duration(<<"Y", d, rest::binary>>) when is_digit(d) # d in ?0..?9
```